### PR TITLE
Authentication and Authorization #40009

### PR DESCRIPTION
**Pull request by shashankbaranwal**
- Dummy JWT, Dummy Person with Roles
- Middleware
- Checking Invalid, Expired, Valid

### CODE REVIEW

File: extraTs/utils/Permissions.ts [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
+        return false
+    }
+    if(temp['all'].includes(role)){
+        return true;
+    }
+    if(!temp[permissionType].includes(role)){
+        return false;
+    }
+    return true;
+}
```
Comment (namanpar) [inline extraTs/utils/Permissions.ts:17]: use es6, proper function naming and define types for parameter

File: extraTs/utils/Validation.ts [hunk 2]
```diff
@@ -1,30 +1,31 @@
...
+        else{
+            b.push(reviewerEmail);
+            countInvalid++;
+        }
+    })
+    console.log("Valid is :"+a);
+    console.log("invalid is :" +b);
+    console.log("Valid: "+countValid+" , Invalid: "+countInvalid);
+};
+
```
Comment (namanpar) [inline extraTs/utils/Validation.ts:31]: use es6, proper function naming and define type for parameters

File: extraTs/utils/helpers.ts [hunk 3]
```diff
@@ -1,10 +1,11 @@
...
-    }
+
+export default function validateEmail(email){
+    const regex = /\w+.\w+@successive.tech$/i;
+    if(regex.test(email))
+    {
+        return true;
+    }
+    else{
+        return false;
```
Comment (namanpar) [inline extraTs/utils/helpers.ts:9]: use es6 and define types

File: src/controllers/trainee/controller.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
+import { Request, Response, NextFunction } from 'express';
+// tslint:disable-next-line: class-name
+class traineeController {
```
Comment (namanpar) [inline src/controllers/trainee/controller.ts:3]: proper ClassName

File: src/controllers/trainee/validation.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
+const config = {
+    get: {
+        skip: {
+            required: false,
+            default: 0,
+            number: true,
+            in: ['query'],
+            errorMessage: 'Skip is invalid'
```
Comment (namanpar) [inline src/controllers/trainee/validation.ts:8]: Is this is a validation file or config file. Use proper naming convention

File: src/libs/routes/authMiddleWare.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+                message: 403
+            });
+        }
+    }
+    catch (err) {
+        next({
+            error: err,
+            message: 'Unauthenticate Access',
+            code: 404
+        });
```
Comment (namanpar) [inline src/libs/routes/authMiddleWare.ts:26]: give name to function and use es6

File: src/libs/routes/errorHandler.ts [hunk 7]
```diff
@@ -1,11 +1,10 @@
 import { Request, Response, NextFunction } from 'express';
 
-export default function errorHandler (err, req: Request, res: Response, next: NextFunction ) {
+export default function errorHandler(err, req: Request, res: Response, next: NextFunction) {
```
Comment (namanpar) [inline src/libs/routes/errorHandler.ts:3]: use es6

File: src/libs/routes/notFoundRoute.ts [hunk 8]
```diff
@@ -1,8 +1,8 @@
...
+    console.log(`first middleWare Working`);
     next({
-        error: "Not Found",
-        code: 500
-    })
-}
\ No newline at end of file
+        err: 'Not Found',
+        code: 404
+    });
```
Comment (namanpar) [inline src/libs/routes/notFoundRoute.ts]: use es6 and name the given function

File: src/libs/routes/permission.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { permissions } from './constant';
+export default function hasPermission(moduleName: string, role: string, permissionType: string): boolean {
```
Comment (namanpar) [inline src/libs/routes/permission.ts:2]: use es6

File: src/libs/routes/validationHandler.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        let request = req[locations][key];
+        const regex = keys.regex;
+        if ((keys.required) && !(request)) {
+            const err = {
+                key: `${key}`,
+                location: `${keys.in}`,
+                errorMessage: `${keys.errorMessage || 'required'}`
+            };
+            errors.push(err);
+        }
```
Comment (namanpar) [inline src/libs/routes/validationHandler.ts:21]: name the function properly

**Review by namanpar — COMMENTED** [review]
Do the requested changes

**Review by namanpar — CHANGES_REQUESTED** [review]
Do the requested changes

File: src/controllers/trainee/validation.ts (hunk 5 above)
Comment (shashankbaranwal) [reply src/controllers/trainee/validation.ts:8]: it is a validation file and inside a constant config is defined as per the guidelines of user story on redmine

File: src/libs/routes/validationHandler.ts (hunk 10 above)
Comment (shashankbaranwal) [reply src/libs/routes/validationHandler.ts:21]: used proper naming function

File: src/libs/routes/permission.ts (hunk 9 above)
Comment (shashankbaranwal) [reply src/libs/routes/permission.ts:2]: Updated the function name and code according to the ES6

File: src/libs/routes/notFoundRoute.ts (hunk 8 above)
Comment (shashankbaranwal) [reply src/libs/routes/notFoundRoute.ts]: Updated the function name and code according to the ES6

File: src/libs/routes/errorHandler.ts (hunk 7 above)
Comment (shashankbaranwal) [reply src/libs/routes/errorHandler.ts:3]: Used ES6 and updated the code

File: src/libs/routes/authMiddleWare.ts (hunk 6 above)
Comment (shashankbaranwal) [reply src/libs/routes/authMiddleWare.ts:26]: Updated the function name according to the ES6

File: src/controllers/trainee/controller.ts (hunk 4 above)
Comment (shashankbaranwal) [reply src/controllers/trainee/controller.ts:3]: Used Proper ClassName please check it

**Review by namanpar — APPROVED** [review]

File: extraTs/utils/Permissions.ts (hunk 1 above)
Comment (shashankbaranwal) [reply extraTs/utils/Permissions.ts]: I am used ES6 and proper naming function as well as defined parameters.

File: extraTs/utils/Validation.ts (hunk 2 above)
Comment (shashankbaranwal) [reply extraTs/utils/Validation.ts]: used ES6 , proper naming function as well as defined parameters.

File: extraTs/utils/helpers.ts (hunk 3 above)
Comment (shashankbaranwal) [reply extraTs/utils/helpers.ts]: Used ES6 and Defined Types in this file.

**Review by namanpar — APPROVED** [review]

**Pull request merged**